lib/evmone/eof.cpp

```diff
@@ -410,46 +411,58 @@ std::variant<EOFValidationError, int32_t> validate_max_stack_height(
 
                 stack_height_required = code_types[func_index].outputs + code_types[fid].inputs -
                                         code_types[fid].outputs;
-                if (stack_heights[i] > stack_height_required)
+
+                // JUMPF to returning function requires exact number of stack items
+                // and is allowed only in constant stack segment.
+                if (stack_height.max > stack_height_required)
                     return EOFValidationError::stack_higher_than_outputs_required;
             }
         }
         else if (opcode == OP_RETF)
         {
             stack_height_required = code_types[func_index].outputs;
-            if (stack_height > code_types[func_index].outputs)
+            // RETF allowed only in constant stack segment
+            if (stack_height.max > stack_height_required)
                 return EOFValidationError::stack_higher_than_outputs_required;
         }
         else if (opcode == OP_DUPN)
             stack_height_required = code[i + 1] + 1;
         else if (opcode == OP_SWAPN)
             stack_height_required = code[i + 1] + 2;
 
-        if (stack_height < stack_height_required)
+        if (stack_height.min < stack_height_required)
             return EOFValidationError::stack_underflow;
 
-        stack_height += stack_height_change;
+        stack_height.min += stack_height_change;
+        stack_height.max += stack_height_change;
 
         // Determine size of immediate, including the special case of RJUMPV.
         const size_t imm_size = (opcode == OP_RJUMPV) ?
                                     (1 + /*count*/ (size_t{code[i + 1]} + 1) * REL_OFFSET_SIZE) :
                                     instr::traits[opcode].immediate_size;
 
-        // Mark immediate locations.
-        std::fill_n(&stack_heights[i + 1], imm_size, LOC_IMMEDIATE);
-
         // Validates the successor instruction and updates its stack height.
-        const auto validate_successor = [&stack_heights, &worklist](size_t successor_offset,
-                                            int32_t expected_stack_height) {
+        const auto validate_successor = [&stack_heights](size_t current_offset,
+                                            size_t successor_offset,
+                                            StackHeightRange expected_stack_height) {
             auto& successor_stack_height = stack_heights[successor_offset];
-            if (successor_stack_height == LOC_UNVISITED)
+            if (successor_offset <= current_offset)  // backwards jump
             {
+                assert(successor_stack_height.min != LOC_UNVISITED);
+                assert(successor_stack_height.max != LOC_UNVISITED);
+                return expected_stack_height.min == successor_stack_height.min &&
+                       expected_stack_height.max == successor_stack_height.max;
```

Comment:
The spec could have been relaxed to
```suggestion
                return expected_stack_height.min >= successor_stack_height.min &&
                       expected_stack_height.max <= successor_stack_height.max;
```
but it was decided against this for simlicity.

Reply:
Leave this comment in the code.

Also, I'd swap the arguments: `successor_stack_height.min <= required_stack_height`.